### ISSUE TITLE
Posts not indexed unless manual sync from admin

### ISSUE THREAD

**Issue by clemorphy** [10up/ElasticPress#3526]
### Describe your question

Hi

I have a lot of content that CAN be indexed, but they are only sent to EP when I manually sync from the admin.
It looks like it is never synced (created or updated) when creating, editing or deleting a post.
Why would the "bulk sync" work correctly while the "one by one" wouldn't?

I have commented out every EP filter I was using (like ep_post_sync_kill or ep_post_query_db_args) thinking maybe that was the issue, but it still only sync when doing it from the "sync now" button...

Any idea on what could interfer?

Same issue on my local env and production.

Thank you

### WordPress Environment ###
wp_version: 6.2.2
home_url: http://localhost
site_url: http://localhost
is_multisite: false
theme: xxxxxxx (1.2.9)
plugins: AuthLDAP (2.4.10), Contact Form 7 (5.6.4), Custom Permalinks (2.4.0), Debug Bar (1.1.4), Debug Bar ElasticPress (2.1.1), Duplicate Page (4.4), ElasticPress (4.6.0), Elasticpress Autosuggest Endpoint (0.3), Enable Media Replace (3.6.3), HappyFiles Pro (1.8.2), LH User Taxonomies (1.60), MCE Table Buttons (3.3), Multiple Roles per User (1), Post Types Order (1.9.3.9), ReOrder Posts within Categories (2.9.2), Revisionize (2.3.4), Theme Switcha (3.2), User Switching (1.7.0), WordPress Importer (0.8.1), WP-Polls (2.75.6), wpDataTables (5.5), WP Last Modified Info (1.4.2), WP Ultimate CSV Importer (6.0.5), et YIKES Simple Taxonomy Ordering (2.3.4)
revisions: all

### Server Environment ###
php_version: 7.4.22
memory_limit: 40M
timeout: 30

## Indexable Content ##
A lot of CPT with a bunch of posts

### Settings ###
host: http://mrc_wsl_elastic:9200
index_prefix: 
language: fr_FR
per_page: 350
network_active: false

### Timeouts ###
request_timeout: 5
index_document_timeout: 15
bulk_request_timeout: 30

### localhost-post-1 ###
health: yellow
status: open
index: localhost-post-1
uuid: 66K6ZRgZSuez74JK7rCgRQ
pri: 5
rep: 1
docs.count: 3366
docs.deleted: 143
store.size: 6.3mb
pri.store.size: 6.3mb
total_fields_limit: 5000

### 2023/07/03 3:39:31   ###
method: WP Dashboard
is_full_sync: No
end_date_time: 2023/07/03 3:39:44  
total_time: 0 heure, 0 minute, 13 secondes
total: 3214
synced: 3316
skipped: 0
failed: 0
errors: array (
)

### Facets ###
active: true
force_inactive: false
match_type: all

### Post Search ###
active: true
decaying_enabled: 0
force_inactive: false
highlight_enabled: 0
highlight_excerpt: 0
highlight_tag: mark
synonyms_editor_mode: simple

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by felipeelia** [comment]
Hi @clemorphy, thanks for opening the issue. Here are some questions:

- Does this happen to all your content or just a subgroup? If so, which are the characteristics of this group? Is it always the same CPT, or some other specificity?
- Are you saving these posts in the Dashboard or programmatically?
- Do you mind sharing your custom code related to ElasticPress?

**Comment by clemorphy** [comment]
Hello @felipeelia 

Sorry for the delay, I was off during the last 2 weeks.

> Does this happen to all your content or just a subgroup? If so, which are the characteristics of this group? Is it always the same CPT, or some other specificity?

It happens to all the content, all CPT.

> Are you saving these posts in the Dashboard or programmatically?

The posts are created, edited, deleted from the dashboard.

> Do you mind sharing your custom code related to ElasticPress?

At the moment, almost all custom code related to EP is commented, except this : 

```
// ElasticPress : ajout du nb d'items par terme dans les facettes
add_filter(
	'ep_facet_widget_term_label',
	function ( $label, $term ) {
		return "{$label} ({$term->count})";
	},
	10,
	2
);

// ElasticPress : pour augmenter le nb de terms à partir duquel un champ de recherche est affiché sur les facet widgets
add_filter(
	'ep_facet_search_threshold',
	function ( $search_threshold ) {
		return 100;
	}
);

```

Thank you for your time.

**Comment by felipeelia** [comment]
Got it, okay. Can you please share the entire report this time (especially the `Indexable Content` section)? It'll probably be easier to upload the txt file instead of copying and pasting it. Thanks!

**Comment by clemorphy** [comment]
Hi @felipeelia 

Here is the full report : [elasticpress-report.txt](https://github.com/10up/ElasticPress/files/12160013/elasticpress-report.txt)

**Comment by felipeelia** [comment]
Hi @clemorphy,

Looking at the report nothing pops up, unfortunately. As this is not reproducible in a fresh install, I'll give you the steps you'll need to further debug it in your codebase:

1. When a post is updated on WordPress, the [wp_update_post()](https://developer.wordpress.org/reference/functions/wp_update_post/) function is called. That ultimately calls wp_insert_post()
2. The wp_insert_post() function calls the [wp_insert_post action](https://developer.wordpress.org/reference/hooks/wp_insert_post/)
3. ElasticPress hooks into that action [here](https://github.com/10up/ElasticPress/blob/4.6.1/includes/classes/Indexable/Post/SyncManager.php#L57)
4. EP's action_sync_on_update method (the one hooked into wp_insert_post) will [add the post to a queue](https://github.com/10up/ElasticPress/blob/4.6.1/includes/classes/Indexable/Post/SyncManager.php#L381)
5. During the shutdown action, EP will [process that queue](https://github.com/10up/ElasticPress/blob/4.6.1/includes/classes/SyncManager.php#L53)

So, you'll need to check if the wp_insert_post is called if our method to add the post to the queue is called, and ultimately, if something is preventing the queue from being processed.

If you’re interested in further assistance, we would be happy to help you. You can visit https://www.elasticpress.io/elasticpress-consulting/ to schedule an appointment for a consultation. Thanks!

**Comment by clemorphy** [comment]
Hi @felipeelia 

Thank you for the checklist. I'm gonna try to debug step by step, and I'll let you know if I eventually find what causes this.

**Comment by clemorphy** [comment]
@felipeelia Thanks to your checklist, it was quite easy to find what is wrong.
Actually the hook to wp_insert_post is never set, because it fails just before.
In the setup method of SyncManager, it never passes the first test:
```
if ( ! Elasticsearch::factory()->get_elasticsearch_version() ) {
	return;
}
```
So it seems it can't retrieve ES version.
I don't think it was the case before, I don't know exactly when it started to fail.
Strangely, the ES version is correctly displayed on the EP settings screen: 

![screenshot-localhost-2023 07 27-16_17_27](https://github.com/10up/ElasticPress/assets/1913472/60415c83-8cd2-4b9f-88a8-4a0bf445017b)

If I comment the test above, everything is syncing just fine.

**Comment by clemorphy** [comment]
Ok I got it.
In the setup method it can't get ES version because the ES host is not defined.
It started to not automatically sync when I moved the host from the settings screen to my source code (in order to add ES_SHIELD and to dynamically set EP_HOST to match my WP environment ).
But when I moved it to my code, I put it in a file that was included in my theme's functions.php.
While it didn't prevent EP to make queries and to bulk index, it was breaking auto sync.
So **EP_HOST and ES_SHIELD constants HAVE TO BE in wp-config.php**, and NOT in functions.php.
Thanks again @felipeelia , I couldn't have debugged this without your last comment.
